### PR TITLE
nushell: load session environment at runtime

### LINKED ISSUE #4313: Make `home.sessionVariables` work with nushell
### Description

Currently i have to set the following:

```nix
  home.sessionVariables = {
    SHELL = "nu";
    TERM = "alacritty";
    EDITOR = "vim";
  };

  programs.nushell.extraEnv = ''
    # FIXME(Krey): This should be handled in `home.sessionVariables`
    $env.SHELL = "nu";
    $env.TERM = "alacritty"
    $env.EDITOR = "vim";

    # Include local binaries
    $env.PATH = ($env.PATH | split row (char esep) | append ($env.HOME | path join '.local/bin'));
  ''; 
```

Because `home.sessionVariables` doesn't export the variables for `nushell` -> Make it work with nushell too?

Expected workflow:

```diff
  home.sessionVariables = {
    SHELL = "nu";
    TERM = "alacritty";
    EDITOR = "vim";
  };

  programs.nushell.extraEnv = ''
-    # FIXME(Krey): This should be handled in `home.sessionVariables`
-    $env.SHELL = "nu";
-    $env.TERM = "alacritty"
-    $env.EDITOR = "vim";
-
    # Include local binaries
    $env.PATH = ($env.PATH | split row (char esep) | append ($env.HOME | path join '.local/bin'));
  ''; 
```

### LINKED ISSUE #4918: bug: nix nushell path environment variable does not have `$USER` and `$HOME`
### Are you following the right branch?

- [X] My Nixpkgs and Home Manager versions are in sync

### Is there an existing issue for this?

- [X] I have searched the existing issues

### Issue description

nushell $env.PATH per-user miss username.
bash path
``` shell
echo "/opt/orbstack-guest/bin-hiprio:/opt/orbstack-guest/data/bin/cmdlinks:/run/wrappers/bin:/home/livenux/.nix-profile/bin:/nix/profile/bin:/home/livenux/.local/state/nix/profile/bin:/etc/profiles/per-user/livenux/bin:/nix/var/nix/profiles/default/bin:/run/current-system/sw/bin:/opt/orbstack-guest/bin" | split row ':'
╭───┬────────────────────────────────────────────╮
│ 0 │ /opt/orbstack-guest/bin-hiprio             │
│ 1 │ /opt/orbstack-guest/data/bin/cmdlinks      │
│ 2 │ /run/wrappers/bin                          │
│ 3 │ /home/livenux/.nix-profile/bin             │
│ 4 │ /nix/profile/bin                           │
│ 5 │ /home/livenux/.local/state/nix/profile/bin │
│ 6 │ /etc/profiles/per-user/livenux/bin         │
│ 7 │ /nix/var/nix/profiles/default/bin          │
│ 8 │ /run/current-system/sw/bin                 │
│ 9 │ /opt/orbstack-guest/bin                    │
╰───┴────────────────────────────────────────────╯
```
nushell path
``` shell
 $env.PATH                                                                                                                                                  01/18/24 01:48:52 AM
╭───┬───────────────────────────────────────╮
│ 0 │ /opt/orbstack-guest/bin-hiprio        │
│ 1 │ /opt/orbstack-guest/data/bin/cmdlinks │
│ 2 │ /run/wrappers/bin                     │
│ 3 │ /.nix-profile/bin                     │
│ 4 │ /nix/profile/bin                      │
│ 5 │ /.local/state/nix/profile/bin         │
│ 6 │ /etc/profiles/per-user//bin           │
│ 7 │ /nix/var/nix/profiles/default/bin     │
│ 8 │ /run/current-system/sw/bin            │
│ 9 │ /opt/orbstack-guest/bin               │
╰───┴───────────────────────────────────────╯
```
flake.nix
``` nix
{
  description = "A template that shows all standard flake outputs";

  inputs = {
    nix.url = "github:NixOS/nix";
    nixpkgs.url = "github:NixOS/nixpkgs/nixos-23.11";

    dwarffs.url = "github:edolstra/dwarffs";

    home-manager = {
      url = "github:nix-community/home-manager/release-23.11";
      inputs.nixpkgs.follows = "nixpkgs";
    };

  };

  outputs = {self, nix, nixpkgs, dwarffs, home-manager, ... }@inputs: {
    nixosConfigurations = {
      nixos = nixpkgs.lib.nixosSystem {
        system = "aarch64-linux";

        modules = [
          ./configuration.nix
          ./users.nix
          dwarffs.nixosModules.dwarffs
          { nixpkgs.overlays = [ nix.overlays.default ]; }

          home-manager.nixosModules.home-manager
          {
            home-manager.useGlobalPkgs = true;
            home-manager.useUserPackages = true;
            home-manager.users.livenux = import ./home-manager.nix;
          }

        ];

      };

    };
  };

}
```

home-mager.nix
``` nix
{ config, pkgs, ... }:
{
  home = {
    username = "livenux";
    homeDirectory = "/home/livenux";
    stateVersion = "23.11";

    packages = [
      pkgs.htop
      pkgs.fortune
    ];
  };

  programs = {
    home-manager.enable = true;

    neovim = {
      enable = true;
      defaultEditor = true;
      viAlias = true;
      vimAlias = true;
    };

    git = {
      enable = true;
      userName = "Livenux";
      userEmail = "livenux@mail.com";
     };
    nushell.enable = true;
  };
}
```

### Maintainer CC

_No response_

### System information

nushell 
``` shell
nix-shell -p nix-info --run "nix-info -m"                                                                                                          
error:
       … while calling the 'import' builtin
         at «string»:1:18:
            1| {...}@args: with import <nixpkgs> args; (pkgs.runCommandCC or pkgs.runCommand) "shell" { buildInputs = [ (nix-info) ]; } ""
             |                  ^

       … while realising the context of a path

       … while calling the 'findFile' builtin
         at «string»:1:25:
            1| {...}@args: with import <nixpkgs> args; (pkgs.runCommandCC or pkgs.runCommand) "shell" { buildInputs = [ (nix-info) ]; } ""
             |                         ^

       error: file 'nixpkgs' was not found in the Nix search path (add it using $NIX_PATH or -I)
```
bash
``` shell
nix-shell -p nix-info --run "nix-info -m"
 - system: `"aarch64-linux"`
 - host os: `Linux 6.5.13-orbstack-00121-ge428743e4e98, NixOS, 23.11 (Tapir), 23.11.20240110.3dc440f`
 - multi-user?: `no`
 - sandbox: `yes`
 - version: `nix-env (Nix) 2.20.0pre20240113_cbd5553`
 - channels(root): `"nixos-23.11.2874.a2fe8d21f667"`
 - nixpkgs: `/nix/var/nix/profiles/per-user/root/channels/nixos`
```

### LINKED ISSUE #6507: bug: nu shell (nushell) with no PATH set on macos
### Are you following the right branch?

- [x] My Nixpkgs and Home Manager versions are in sync

### Is there an existing issue for this?

- [x] I have searched the existing issues

### Issue description

Using nushell as default shell results in PATH not being set correctly so totally unusable.
using home-manager 24.11 vs nix 2.24.12 vs nix-darwin 25.05 
using 'nu' from the zsh shell works.
Image

using nu from shell default chsh -s .../bin/nu dont' work path
Image

same issues: https://github.com/nix-community/NixOS-WSL/issues/399

### Maintainer CC

_No response_

### System information

```markdown
- system: `"aarch64-darwin"`
 - host os: `Darwin 24.3.0, macOS 15.3`
 - multi-user?: `yes`
 - sandbox: `no`
 - version: `nix-env (Nix) 2.24.12`
 - channels(root): `"nixpkgs"`
 - nixpkgs: `/nix/store/7770rp8l36gr0ayj230ww1idjpl3rkwf-source`
```

### LINKED ISSUE #6901: NuShell as default shell with Home Manager on Darwin (macOS)?
I'm trying to use **NuShell** as my default shell on **macOS with Home Manager and Nix-Darwin**, but I'm running into an issue where my **Home Manager/Nix environment doesn't load** when NuShell is the login shell.

## What I did

- My NuShell binary is located at:  
  `/etc/profiles/per-user/phuc/bin/nu`

- I added this path to `/etc/shells`:
  ```bash
  sudo echo /etc/profiles/per-user/phuc/bin/nu >> /etc/shells
```
- then:
```bash
chsh -s /etc/profiles/per-user/phuc/bin/nu
```
## ❌ What happens
	•	After restarting the terminal, NuShell runs correctly, but My Nix environment and Home Manager configuration do not load.
	•	Only `zsh` function normally.
##❓ Question
How can I properly set NuShell as the default shell on `Nix-Darwin` or `Home Manager`, and still retain the full Nix environment and Home Manager configuration?

Any guidance or working setup examples would be greatly appreciated!

### PR DESCRIPTION
**Pull request by LisaScheers**
## What changed

- Generate a Nushell `env.nu` loader when a system or Home Manager session
  environment needs to be applied.
- At shell startup, source the OS-generated `setEnvironment` script when Home Manager is used as a NixOS or nix-darwin module, followed by `hm-session-vars.sh`.
- Capture only variables changed by those scripts and load them into Nushell, preserving `PATH` as a `list<string>`.
- Keep the existing source guards and load the generated environment before the user's `envFile`/`extraEnv`, so inherited or user-specific values keep their expected precedence.
- Document the new `env.nu` ownership and migration path in the 26.11 release
  notes.

## Why

Nushell does not source the POSIX environment scripts used by Bash and Zsh. As a result, a directly launched Nushell login shell can miss the Nix profile `PATH`, Home Manager session variables, and expansions involving runtime values such as `$HOME` and `$USER`.

Evaluating these scripts at build time captures sandbox values instead of the user's login environment. This implementation evaluates them at shell startup and transfers the resulting environment through a NUL-delimited representation, which also handles values containing newlines.

This builds on the investigation and experiments in #6941 while avoiding its sandbox-environment limitation.

## User impact

Nushell can be launched directly instead of through Zsh or Bash while receiving the system and Home Manager environment. `home.sessionVariables`, `home.sessionPath`, POSIX default/alternate expansions, and NixOS/nix-darwin profile paths are applied consistently.

Fixes #4313
Fixes #4918
Fixes #6507
Addresses #6568
Fixes #6901

## Validation

- Added a runtime test that launches Nushell from an intentionally minimal environment and verifies OS variables, `$HOME`/`$USER` expansion, POSIX default/alternate expansion, `home.sessionPath`, list-form `PATH`, and user `env.nu` precedence.
- Ran the full test suite with `nix build .#test-all`.
- Built all 13 Nushell-focused Home Manager module tests on `aarch64-darwin`.
- Evaluated the new test derivation for `x86_64-linux`.
- Rendered the HTML manual with `nix build .#docs-html`.
- Ran the repository formatter, the 80-column check for added lines, and
  `git diff --check`.

### CODE REVIEW

**Comment by iniw** [comment]
Note that this PR still doesn't you to seamlessly set nushell as the login shell on macOS if you want the config to be stored in XDG directories. If `XDG_CONFIG_HOME` is not set _during startup_, nushell will look for the config file in `~/Library/Application Support/nushell`, which is not where home manager has placed it if you have the XDG-related options enabled. I'm not sure what the best way to fix it is 🤷

**Comment by LisaScheers** [comment]
the real solution is to let the xdg module write a plist with something like
```plist
<!-- ~/Library/LaunchAgents/setenv.XDG_CONFIG_HOME.plist -->
<?xml version="1.0" encoding="UTF-8"?>
<!DOCTYPE plist PUBLIC "-//Apple//DTD PLIST 1.0//EN" "http://www.apple.com/DTDs/PropertyList-1.0.dtd">
  <plist version="1.0">
  <dict>
  <key>Label</key>
  <string>setenv.XDG_CONFIG_HOME</string>
  <key>ProgramArguments</key>
  <array>
    <string>/bin/launchctl</string>
    <string>setenv</string>
    <string>XDG_CONFIG_HOME</string>
    <string>{USER_HOME}/.config</string>
  </array>
  <key>RunAtLoad</key>
  <true/>
</dict>
</plist>
```
to `~/Library/LaunchAgents/setenv.XDG_CONFIG_HOME.plist`

this would set it for all applications

**Comment by tb148** [comment]
I wonder if the "OS environment" part of this PR can be applied to the new nushell nixos module at https://github.com/NixOS/nixpkgs/blob/nixos-unstable/nixos/modules/programs/nushell.nix by using vendor autoload dirs

File: modules/programs/nushell.nix [hunk 1]
```diff
@@ -9,6 +10,102 @@ let
...
+      captureEnvironment = ''
+        ${lib.getExe' pkgs.coreutils "env"} -0
+        printf '\0'
+        ${sourceSystemEnvironment}
+        . ${lib.escapeShellArg homeSessionVariablesFile} >/dev/null
+        ${lib.getExe' pkgs.coreutils "env"} -0
+      '';
+      shouldLoadSystemEnvironment = lib.optionalString hasSystemEnvironment ''
+        or (
+          "__NIXOS_SET_ENVIRONMENT_DONE" not-in $env
```
Comment (khaneliman) [inline modules/programs/nushell.nix:52]: issue (blocking): treat empty system markers as not initialized

The skip condition checks only whether each marker exists, but nix-darwin's [tmux wrapper sets `__NIX_DARWIN_SET_ENVIRONMENT_DONE` to `""`](https://github.com/nix-darwin/nix-darwin/blob/4cff07de74b50e64bdd68cd4e722ab5b6b35ee48/modules/programs/tmux.nix#L17-L22) so child shells reload the system environment; its [Zsh loader therefore checks `-z`](https://github.com/nix-darwin/nix-darwin/blob/4cff07de74b50e64bdd68cd4e722ab5b6b35ee48/modules/programs/zsh/default.nix#L158-L161). Mirror that value check here with optional access:

```nu
or (
  ($env.__NIXOS_SET_ENVIRONMENT_DONE? | is-empty)
  and ($env.__NIX_DARWIN_SET_ENVIRONMENT_DONE? | is-empty)
)
```

Please also test `__HM_SESS_VARS_SOURCED=1` together with an empty Darwin marker; that case must reload `system.build.setEnvironment`.

File: modules/programs/nushell.nix [hunk 2]
```diff
@@ -9,6 +10,102 @@ let
...
+          | first
+          | split row (char nul)
+          | compact --empty
+        )
+        let changed = (
+          $sections
+          | last
+          | split row (char nul)
+          | compact --empty
+          | where { |entry| $entry not-in $before }
```
Comment (khaneliman) [inline modules/programs/nushell.nix:94]: issue (blocking): propagate variables removed by the sourced scripts

Comparing complete `KEY=value` strings detects additions and changes, but it cannot represent removal. Parse both snapshots into records keyed by variable name, apply added or changed values with `load-env`, then remove names missing from the post-source record:

```nu
for name in ($before | columns) {
  if $name not-in $after and $name not-in ["_" "_AST_FEATURES" "SHLVL"] {
    hide-env --ignore-errors $name
  }
}
```

Home Manager's [Nushell direnv integration uses this same dynamic `hide-env` pattern](https://github.com/nix-community/home-manager/blob/165228b0efefc3e635e5174020c40ea64271dc25/modules/programs/direnv.nix#L153-L160), because `load-env` does not remove absent keys. Add a test that inherits a variable, unsets it from the sourced environment script, and verifies it is absent in Nushell afterward.
